src/kOS/Serialization/ConfigNodeFormatter.cs

```diff
@@ -0,0 +1,184 @@
+            ConfigNode innerNode = new ConfigNode();
+            innerNode.name = ListKey;
+
+            var indexedList = list.Select((value, index) => new { Value = value, Index = index.ToString() });
+
+            foreach (var item in indexedList) {
+                HandleValue(innerNode, item.Index, item.Value);
+            }
+
+            configNode.AddNode(innerNode);
+
+            return configNode;
+        }
+
+        private ConfigNode ScalarToConfigNode(string name, string type, string value)
+        {
+            ConfigNode configNode = new ConfigNode();
+            configNode.name = name;
+
+            // ConfigNode stores all values as strings so we have to store non-string values as nodes, otherwise
+            // the type information would be lost
+            ConfigNode innerNode = new ConfigNode();
+            innerNode.name = type;
+
+            innerNode.AddValue(ValueKey, value);
+
+            configNode.AddNode(innerNode);
+
+            return configNode;
+        }
+
+        private ConfigNode ToConfigNode(string name, Dump dump)
+        {
+            ConfigNode configNode = new ConfigNode();
+
+            configNode.name = name;
+
+            foreach (var entry in dump) {
+                HandleValue(configNode, entry.Key.ToString(), entry.Value);
+            }
+
+            return configNode;
+        }
+
+        private void HandleValue(ConfigNode configNode, string key, object value)
+        {
+            if (value is Dump)
+            {
+                configNode.AddNode(key, ToConfigNode(key, value as Dump));
+            } else if (value is List<object>)
+            {
+                configNode.AddNode(key, ToConfigNode(key, value as List<object>));
+            } else if (value is int) {
+                configNode.AddNode(key, ScalarToConfigNode(key, ScalarKey, Convert.ToString(value)));
+            } else if (value is double) {
+                configNode.AddNode(key, ScalarToConfigNode(key, ScalarKey, ((double)value).ToString("R")));
+
+            } else if (value is bool) {
+                configNode.AddNode(key, ScalarToConfigNode(key, BooleanKey, Convert.ToString(value)));
+            } else
+            {
+                configNode.AddValue(key, value);
+            }
+        }
+
+        private List<object> ListFromConfigNode(ConfigNode configNode)
+        {
+            List<object> result = new List<object>();
+
+            int i = 0;
+
+            while (true) {
+                if (configNode.HasValue(i.ToString()))
+                {
+                    result.Add(configNode.GetValue(i.ToString()));
+                } else if (configNode.HasNode(i.ToString()))
+                {
+                    result.Add(ObjectFromConfigNode(configNode.GetNode(i.ToString())));
+                } else
+                {
+                    break;
+                }
+
+                i++;
+            }
```

Comment:
(I also think that if you definitely do have an index variable that gets initialized and counts up per iteration, as you have here, then the 3-part for loop makes more sense than "while", but I know I'm not in the majority on that opinion.)